### PR TITLE
Use Litani for running proofs

### PR DESCRIPTION
**Pull request by karkhaz**
This changes Makefile.common to invoke Litani when building proofs. Litani is a tool hosted at [awslabs/aws-build-accumulator](https://github.com/awslabs/aws-build-accumulator).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by karkhaz** [comment]
I wonder whether a fairly vanilla version of the `run-all-proofs.py` script should also be included in this PR. It would be copied into the proof directory rather than symlinked in, since we would expect proof-writers to change it fairly quickly.

File: template-for-repository/proofs/Makefile.common [hunk 1]
```diff
@@ -246,45 +247,10 @@ ARPA_COMP_CMDS ?= $(ARPA_BLDDIR)/compile_commands.json
...
-C_FROM_GOTO = $(patsubst $(GOTODIR)%,$(SRCDIR)%,$(1:.goto=.c))
-GOTO_FROM_C = $(patsubst $(SRCDIR)%,$(GOTODIR)%,$(1:.c=.goto))
-LOG_FROM_ENTRY = $(LOGDIR)/$(notdir $(1:.goto=.txt))
-LOG_FROM_GOTO = $(patsubst $(GOTODIR)%,$(LOGDIR)%,$(1:.goto=.txt))
-
 ################################################################
 # Set C compiler defines
 
-CBMCFLAGS +=  --object-bits $(CBMC_OBJECT_BITS)
+CBMCFLAGS += --object-bits $(CBMC_OBJECT_BITS) --flush
```
Comment (SaswatPadhi) [inline template-for-repository/proofs/Makefile.common:253]: [Unrelated to changes in PR]

May be use this section only for compiler defines, and move line 253 next to line 113 to keep it together with other `CBMCFLAGS` stuff?

**Review by SaswatPadhi — COMMENTED** [review]
LGTM. Just one comment on `CBMCFLAGS`:

File: template-for-repository/proofs/Makefile-project-defines [hunk 2]
```diff
@@ -13,6 +13,18 @@
...
+# "s2n" or "Amazon FreeRTOS". For projects with multiple proof roots,
+# this may be overridden on the command-line to Make, for example
+#
+# 	  make PROJECT_NAME="FreeRTOS MQTT" report
+#
+# PROJECT_NAME =
+
+# Path to litani executable, relative to the root of the repository.
+# This applies even for projects with multiple proof roots.
+# LITANI =
```
Comment (karkhaz) [inline template-for-repository/proofs/Makefile-project-defines:26]: I wonder whether I should modify the `setup.py` script to prompt the user for the Litani directory, and write it next to `SRCDIR`. The setup script could optionally add the submodule to the location user specified, if it didn't exist. Thoughts?

Comment (markrtuttle) [reply template-for-repository/proofs/Makefile-project-defines:26]: It should prompt for those things that are required for make report to run.

File: template-for-repository/proofs/Makefile.common [hunk 3]
```diff
@@ -37,6 +37,7 @@
 #
 #         HARNESS_FILE=<HARNESS_FILE>
 #         HARNESS_ENTRY=<HARNESS_ENTRY>
+#         PROOF_UID=<PROOF_UID>
```
Comment (markrtuttle) [inline template-for-repository/proofs/Makefile.common:40]: Where does this come from?  What is the proof_uid?  Should the set up script prompt for?  I kind of hope not, but the script should set this.

File: template-for-repository/proofs/Makefile.common [hunk 4]
```diff
@@ -158,6 +159,8 @@ CHECKFLAGS += $(CBMC_FLAG_SIGNED_OVERFLOW_CHECK)
 CHECKFLAGS += $(CBMC_FLAG_UNDEFINED_SHIFT_CHECK)
 CHECKFLAGS += $(CBMC_FLAG_UNSIGNED_OVERFLOW_CHECK)
 
+CHECKFLAGS += --unwinding-assertions
```
Comment (markrtuttle) [inline template-for-repository/proofs/Makefile.common:162]: Why not use the same defined-variable approach here?  Yes, everyone should use it, but we want to encourage best practice without forcing anything.

File: template-for-repository/proofs/Makefile.common [hunk 5]
```diff
@@ -337,54 +305,109 @@ CBMC_REMOVE_FUNCTION_BODY := $(patsubst %,--remove-function-body %, $(REMOVE_FUN
 
 # Compile project sources
 $(PROJECT_GOTO)1.goto: $(PROJECT_SOURCES)
-	mkdir -p $(dir $@)
-	mkdir -p $(dir $(call LOG_FROM_ENTRY,$@))
-	$(call DO_GOTO_CC,--export-file-local-symbols $(COMPILE_FLAGS) $(INCLUDES) $(DEFINES),$^,$@)
+	$(LITANI) add-job \
```
Comment (markrtuttle) [inline template-for-repository/proofs/Makefile.common:308]: In all of these litani invocations, didn't we say we should collect all stdout and stderr and log it?

File: template-for-repository/proofs/Makefile.common [hunk 6]
```diff
@@ -402,31 +425,77 @@ arpa: $(ARPA_BLDDIR)
...
 $(LOGDIR)/result.txt: $(HARNESS_GOTO).goto
-	$(call DO_CBMC,$(CBMCFLAGS) $(CHECKFLAGS) --unwinding-assertions --trace,$<,$@)
+	$(LITANI) add-job \
+	  --command \
+	    '$(CBMC) $(CBMC_VERBOSITY) $(CBMCFLAGS) $(CHECKFLAGS) --trace $<' \
+	  --inputs $^ \
+	  --outputs $@ \
+	  --pipeline-name "$(PROOF_UID)" \
+	  --ci-stage test \
+	  --stdout-file $@ \
```
Comment (markrtuttle) [inline template-for-repository/proofs/Makefile.common:435]: Again, in all of these litani invocations, don't we want to collect and interleave all stdout stderr output?

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
There are two unresolved comments from #35, but otherwise this PR LGTM.

https://github.com/awslabs/aws-templates-for-cbmc-proofs/pull/35#discussion_r506582962

https://github.com/awslabs/aws-templates-for-cbmc-proofs/pull/35#discussion_r505956984

**Review by markrtuttle — CHANGES_REQUESTED** [review]
I think we need to change the cbmc-batch.yaml in the proof template directory to be named cbmc-proof.txt.

File: template-for-repository/proofs/Makefile.common (hunk 3 above)
Comment (karkhaz) [reply template-for-repository/proofs/Makefile.common:40]: This is explained in the template-for-proof Makefile:
```
# This should be a unique identifier for this proof, and will appear on the
# Litani dashboard. It can be human-readable and contain spaces if you wish.
PROOF_UID = <__FUNCTION_NAME__>
```

File: template-for-repository/proofs/Makefile.common (hunk 5 above)
Comment (karkhaz) [reply template-for-repository/proofs/Makefile.common:308]: done

File: template-for-repository/proofs/Makefile.common (hunk 6 above)
Comment (karkhaz) [reply template-for-repository/proofs/Makefile.common:449]: done

File: template-for-repository/proofs/Makefile-project-defines (hunk 2 above)
Comment (karkhaz) [reply template-for-repository/proofs/Makefile-project-defines]: done, thanks

File: template-for-repository/proofs/Makefile.common (hunk 4 above)
Comment (karkhaz) [reply template-for-repository/proofs/Makefile.common]: done

**Review by markrtuttle — APPROVED** [review]
